core/lib/contract_verifier/src/resolver.rs

```diff
@@ -0,0 +1,220 @@
+    ///
+    /// # Errors
+    ///
+    /// Returned errors are assumed to be fatal.
+    async fn supported_versions(&self) -> anyhow::Result<SupportedCompilerVersions>;
+
+    /// Resolves a `zksolc` compiler.
+    async fn resolve_solc(
+        &self,
+        versions: &CompilerVersions,
+    ) -> Result<Box<dyn Compiler<ZkSolcInput>>, ContractVerifierError>;
+
+    /// Resolves a `zkvyper` compiler.
+    async fn resolve_vyper(
+        &self,
+        versions: &CompilerVersions,
+    ) -> Result<Box<dyn Compiler<ZkVyperInput>>, ContractVerifierError>;
+}
+
+/// Encapsulates a one-off compilation process.
+#[async_trait]
+pub(crate) trait Compiler<In>: Send + fmt::Debug {
+    /// Performs compilation.
+    async fn compile(
+        self: Box<Self>,
+        input: In,
+    ) -> Result<CompilationArtifacts, ContractVerifierError>;
+}
+
+/// Default [`CompilerResolver`] using pre-downloaded compilers in the `/etc` subdirectories (relative to the workspace).
+#[derive(Debug)]
+pub(crate) struct EnvCompilerResolver {
+    home_dir: PathBuf,
+}
+
+impl Default for EnvCompilerResolver {
+    fn default() -> Self {
+        Self {
+            home_dir: Workspace::locate().core(),
+        }
+    }
+}
+
+impl EnvCompilerResolver {
+    async fn read_dir(&self, dir: &str) -> anyhow::Result<Vec<String>> {
+        let mut dir_entries = fs::read_dir(self.home_dir.join(dir))
+            .await
+            .context("failed reading dir")?;
+        let mut versions = vec![];
+        while let Some(entry) = dir_entries.next_entry().await? {
+            let Ok(file_type) = entry.file_type().await else {
+                continue;
+            };
+            if file_type.is_dir() {
+                if let Ok(name) = entry.file_name().into_string() {
+                    versions.push(name);
+                }
+            }
+        }
+        Ok(versions)
+    }
+}
+
+#[async_trait]
+impl CompilerResolver for EnvCompilerResolver {
+    async fn supported_versions(&self) -> anyhow::Result<SupportedCompilerVersions> {
+        Ok(SupportedCompilerVersions {
+            solc: self
+                .read_dir("etc/solc-bin")
+                .await
+                .context("failed reading solc dir")?,
+            zksolc: self
+                .read_dir("etc/zksolc-bin")
+                .await
+                .context("failed reading zksolc dir")?,
+            vyper: self
+                .read_dir("etc/vyper-bin")
+                .await
+                .context("failed reading vyper dir")?,
+            zkvyper: self
+                .read_dir("etc/zkvyper-bin")
+                .await
+                .context("failed reading zkvyper dir")?,
+        })
+    }
+
+    async fn resolve_solc(
+        &self,
+        versions: &CompilerVersions,
+    ) -> Result<Box<dyn Compiler<ZkSolcInput>>, ContractVerifierError> {
+        let zksolc_version = versions.zk_compiler_version();
+        let zksolc_path = self
```

Comment:
Nit: This part feels overly repeated. Could we encapsulate it to something like `CompilerPaths::try_detect(&self.home_dir, Compiler::ZkSolc).await?` or `Compiler::ZkSolc::try_detect(&self.home_dir).await?` ?